pkg/v2/tkr/webhook/cluster/vsphere-template-resolver/templateresolver/resolver_impl.go

```diff
@@ -0,0 +1,179 @@
+	if err != nil {
+		r.Log.Error(err, "failed to get VSphereVirtualMachine images from VC client")
+		return Result{UsefulErrorMessage: err.Error()}
+	}
+
+	// Find VMs with matching ova versions, and update relevant fields.
+	return updateTemplateDetailsInVMs(vcVMs, query)
+}
+
+// combine returns a new map containing a union of all pairs in the input maps.
+func combine(m1 map[TemplateQuery]struct{}, m2 map[TemplateQuery]struct{}) map[TemplateQuery]struct{} {
+	combined := map[TemplateQuery]struct{}{}
+	for k, v := range m1 {
+		combined[k] = v
+	}
+	for k, v := range m2 {
+		combined[k] = v
+	}
+	return combined
+}
+
+// getVirtualMachineTemplateForOVAs queries VC and retrieves all the `VSphereVirtualMachine` entries.
+func getVirtualMachineTemplateForOVAs(dc string, vcClient vc.Client) ([]*types.VSphereVirtualMachine, error) {
+	// We need DC MOID to query VMs.
+	dcMOID, err := vcClient.FindDataCenter(context.TODO(), dc)
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to get the datacenter MOID")
+	}
+
+	// Get all vcVMs.
+	vcVMs, err := vcClient.GetVirtualMachineImages(context.TODO(), dcMOID)
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to get K8s VM templates")
+	}
+
+	return vcVMs, nil
+}
+
+// updateTemplateDetailsInVMs: iterates over the list of VSphereVirtualMachine and finds the matching query associated with it.
+// Returns a result containing the template path and MOID for each query.
+func updateTemplateDetailsInVMs(vcVMs []*types.VSphereVirtualMachine, query Query) Result {
+	cpQueries := query.ControlPlane
+	mdQueries := query.MachineDeployments
+
+	cpResult := OVATemplateResult{}
+	mdResult := OVATemplateResult{}
+
+	// Combining the CP and MD queries and results so it is easy to keep track of what is fulfilled and what is not.
+	combinedQueries := combine(cpQueries, mdQueries)
+
+	nonTemplateVMs := map[TemplateQuery][]string{}
+	for _, vm := range vcVMs {
+		query := TemplateQuery{
+			OVAVersion: vm.OVAVersion,
+			OSInfo: v1alpha3.OSInfo{
+				Arch:    vm.DistroArch,
+				Name:    vm.DistroName,
+				Version: vm.DistroVersion,
+			},
+		}
+
+		if _, ok := combinedQueries[query]; !ok {
+			// No query that matches this, or query already fulfilled in previous iteration.
+			continue
+
+		}
+
+		if !vm.IsTemplate {
+			// Although VM matches query, this is not a template VM.
+			// Collect information for ppopulating a useful error message.
+			nonTemplateVMs[query] = append(nonTemplateVMs[query], vm.Name)
+			continue
+		}
+
+		// Query exists, time to create a template result for it.
+		templateResult := &TemplateResult{
+			TemplatePath: vm.Name,
+			TemplateMOID: vm.Moid,
+		}
+
+		if _, ok := cpQueries[query]; ok {
+			cpResult[query] = templateResult
+		}
+
+		if _, ok := mdQueries[query]; ok {
+			mdResult[query] = templateResult
+		}
+
+		// Empty the info collected for useful error message
+		delete(nonTemplateVMs, query)
+		// Empty the entry from input so the missing values can be checked.
```

Comment:
This a good question. I'm going to add an explanation to make it more clear.

```go
// Empty the map that collects info for useful error message since template VM is found.
// This will be a no-op unless the map was populated in a previous iteration due to a non-template VM match being found.
// Now that we have found a template VM that matches the query, there is no need to have the entry in this map.
```